airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,155 @@
+            interval of deleting serialized DAGs in DB when the files are deleted, suggest
+            to use a smaller interval such as 60
+
+    It is used by webserver to load dagbags when dagcached=True. Because reading from
+    database is lightweight compared to importing from files, it solves the webserver
+    scalability issue.
+    """
+    __tablename__ = 'serialized_dag'
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    fileloc = Column(String(2000))
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer)
+    data = Column(Text)
+    last_updated = Column(UtcDateTime)
+
+    __table_args__ = (
+        Index('idx_fileloc_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, dag):
+        from airflow.dag.serialization import Serialization
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
+        self.data = Serialization.to_json(dag)
+        self.last_updated = timezone.utcnow()
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        return int(0xFFFF & int(
+            hashlib.sha1(full_filepath.encode('utf-8')).hexdigest(), 16))
+
+    @classmethod
+    def write_dag(cls, dag: 'DAG', min_update_interval: Optional[int] = None):
+        """Serializes a DAG and writes it into database.
+
+        :param dag: a DAG to be written into database
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        """
+        with db.create_session() as session:
+            if min_update_interval is not None:
+                result = session.query(cls.last_updated).filter(
+                    cls.dag_id == dag.dag_id).first()
+                if result is not None and (
+                        timezone.utcnow() - result.last_updated).total_seconds() < min_update_interval:
+                    return
+            session.merge(cls(dag))
+
+    @classmethod
+    def read_all_dags(cls) -> Dict[str, 'SerializedDAG']:
+        """Reads all DAGs in serialized_dag table.
+
+        :returns: a dict of DAGs read from database
+        """
+        from airflow.dag.serialization import Serialization
+
+        with db.create_session() as session:
+            serialized_dags = session.query(cls.dag_id, cls.data).all()
+
+        dags = {}
+        for dag_id, data in serialized_dags:
+            dag = Serialization.from_json(data)  # type: Any
+            # Sanity check.
+            if dag.dag_id == dag_id:
```

Comment:
If we are checking this we should do _something_ when it's false.

Reply:
I will add a warning note.

Reply:
Done in https://github.com/apache/airflow/pull/5743/commits/a70027b5dcbab9a65224d62f818c20c6f5c9ad5c